### ISSUE TITLE
Verifier not displayed on receiving end if wormhole code provided too late

### ISSUE THREAD

**Issue by jpolnetpl** [warner/magic-wormhole#349]
For transfers that requested verification (`--verify` option on both ends, `send` and `receive`), the receiving end has very little time to provide wormhole code to be able to see the "Verifier" before transfer.

If one fails to do so quickly enough, only "Key established, waiting for confirmation..." message is displayed.

This happens always when the code isn't provided as a command line argument to `wormhole receive -v`, or it isn't entered very soon (one second?) after the "Enter receive wormhole code:" prompt.

And it looks like the "waiting for confirmation..." message is displayed forever, unless the sender decides to abort the transfer (or accept the verifier even though the receiver didn't have the chance to confirm it). After that the "Verifier" is actually displayed on `wormhole receive` side (which is unfortunately too late).

I believe this behaviour doesn't manifest with version 0.9.1 on receiving end (sender may use 0.9.1 or 0.11.2, it doesn't change anything). I haven't tested any other versions.

**Comment by jpolnetpl** [comment]
I've done some more tests. It seems the problem occurs only when tab-completion is used to enter the wormhole code on receiving end. When I entered (even slowly) the whole code or pasted it, the verifier appears. But still, this error doesn't occur in version 0.9.1 and it does in 0.11.2.

**Comment by warner** [comment]
Ooh, weird. Thanks for the catch, and for the detailed testing.. that helps me a lot.

I've confirmed the bug locally.. I can't figure out why it depends upon tab-completion on the receiver, but there's certainly a race condition in the sender between displaying/waiting for the `ok? (yes/no)` message and sending the VERSIONS message to the receiver. The receiver does not display the `Verifier: hehex` message until it gets that VERSIONS message, so there's definitely a case there which matches the bug.

Lemme write down the exact sequence of events to get it clear in my mind (and to document it for the future):

## Sequence of Events

The sequence of events on the sending side is:

* S0: we generate the wormhole code
* S1: we send the PAKE message
* S2: we receive the receiver's PAKE message, compute an "unverified" session key, and send the VERSIONS message
* S3: we receive the receiver's VERSIONS message and "verify" it as matching our own session key
* S4: if `--verify` was requested, we display a hash of the session key as `Verifier hexhex. ok? (yes/no)`, and wait for a response by the user. If they say `no`, we send an error message to the far side and shut down. This uses a blocking python stdlib `input()` function.
* S5: we send an "offer" with the filename we want to send
* S6: we wait for a "permission" response, then start sending

The sequence of events on the receiving side is:

* R0: we learn the full wormhole code
* R1: we send the PAKE message
* R2: we receive the sender's PAKE message, compute an "unverified" session key, and send VERSIONS
* R3: we receive the sender's VERSIONS message and "verify" it as matching our own session key
* R4: if `--verify` was used, we display `Verifier (hexhex).` with a hash of the session key
* R5: we receive the sender's "offer", telling us the filename they want to send
* R6: we wait for the user to approve the transfer
* R7: we send a "permission" response that tells the sender to start sending
* R8: we receive and store the file

## Timeouts

On the receiver, if R2 doesn't arrive within one second of R1, we display `Waiting for sender...`. In the usual case, where the sender created their code ages ago (assuming it takes a minute for one person to speak the code to the other, and for the second person to type it in), the sender's initial PAKE message is already on the server, so to get from R1 to R2 only requires a network roundtrip to the server. This message will be printed if you got the number part of the wormhole code wrong (so the receiver is looking in the wrong mailbox), or if you're using the "offline codes" feature (where the two people agree on a code without their computers and then go back home and type them in) and the receiver types in their code first.

On the receiver (again), if R3 doesn't arrive within one second of R2, we display `Key established, waiting for confirmation`. Normally this should be pretty quick: we need to make a roundtrip (through the server) to the sender, and the sender has to do some math. But if the sender has quit before the receiver types in their code, you'll see this message, and the receiver would wait forever (or until you interrupt it). The "waiting for confirmation" in that message is unrelated to the `--verify` hex string.

On the sender, if S3 doesn't arrive within one second of S2 arriving, it prints `Key established, waiting for confirmation...`. This should only happen if the receiver did R1 but not R2, which could happen if you kill the process at exactly the right time (within a single roundtrip, so maybe 150 milliseconds), or if you're using offline codes and the receiver typed theirs in first and then the receiver quit.

## --verify

Neither side shows the verifier until we know we're talking to someone with the same session key. If an MitM attacker guesses the code wrong (with probability 99.99847%, 65535-out-of-65536), and allows the connection to proceed far enough to send the VERSIONS message, then both sides will abort at step S3/R3 with a WrongPasswordError. If they guessed the code right (probability 0.00153%, 1-out-of-65536) then `--verify` is your only hope.

The idea was to not ask the user to compare something that we know can't possibly be correct. Although to achieve that, we hold off from displaying the verifier until we know it *might* be correct, which is not exactly the same thing.

The sending user sees the verifier in step S4, and that's also where they get to abandon the transfer if they decide it doesn't match. The receiving user sees the verifier in step R4, and if it doesn't match what the sender is seeing, step R6 is where they reject the transfer.

## hypothesis

If the receiver sends PAKE and VERSIONS fast enough (i.e. the delay between R1 and R2 is short enough that the server delivers both in the same Websocket bundle), then the sender will process both on the same reactor turn. In that turn, the sender queues the VERSIONS message for transmission in S2, but then enters the blocking `input()` call before returning control to the reactor. The sender's VERSIONS message therefore doesn't get transmitted to the server, so the receiver never gets it, so the receiver will be waiting for R3 long enough to display the `Key established, waiting for confirmation` message and never shows the Verifier.

## questions (for myself)

* why does tab-completion make a difference? The timing might be affected. The receiver cannot compute their PAKE message until the code is known, and cannot compute their VERSIONS message until the sender's PAKE message is received. Are we blocking inside `readline()` and not receiving PAKE until it completes? I don't think so, since we're tab-completing on the nameplate, and that requires network roundtrips to the server.
* maybe we could just add a short delay (`reactor.callLater(0.5)`) before calling `input()` to let the VERSIONS get sent
* should we add an `ok? (yes/no)` to the receiver too? we're currently only having the sender provide confirmation (I was focussed on the confidentiality of the sender's file). Do we have enough messages to allow the sender to react properly to the receiver declining confirmation? Can we re-use the message sent when the receiver declines permission?

**Comment by warner** [comment]
A one-millisecond delay seems to be sufficient to let the receiver show the verifier when using tab-completion, but I'm going to keep digging and understand why tab-completion makes a difference.

```
% git diff
diff --git a/src/wormhole/cli/cmd_send.py b/src/wormhole/cli/cmd_send.py
index 8326717..09b2b44 100644
--- a/src/wormhole/cli/cmd_send.py
+++ b/src/wormhole/cli/cmd_send.py
@@ -12,7 +12,7 @@ import six
 from humanize import naturalsize
 from tqdm import tqdm
 from twisted.internet import reactor
-from twisted.internet.defer import inlineCallbacks, returnValue
+from twisted.internet.defer import inlineCallbacks, returnValue, Deferred
 from twisted.protocols import basic
 from twisted.python import log
 from wormhole import __version__, create
@@ -160,6 +160,9 @@ class Sender:
                 notify.cancel()

         if args.verify:
+            d = Deferred()
+            reactor.callLater(0.001, d.callback, None)
+            yield d
             self._check_verifier(w,
                                  verifier_bytes)  # blocks, can TransferError
```

**Comment by warner** [comment]
In the tab-completion case, the receiver sends RELEASE (when the user hits TAB after the nameplate). Then later, when the user finishes entering the code and hits ENTER, the receiver sends PAKE and ADD(VERSION) in a single turn.

In the non-tab-completion case, nothing happens until ENTER, at which point the receiver sends PAKE and RELEASE and ADD(VERSION) in a single turn.

So the sender is receiving RELEASE before or after the PAKE/ADD(VERSION), but I'm not yet sure how that affects things.

**Comment by warner** [comment]
In the tab-completion case, the mailbox server gets the receiver's ADD(PAKE) and ADD(VERSION) in a single turn, and thus sends both PAKE and VERSION to the sender in the same turn, so the websocket/TCP layer is free to merge them into a single IP frame, so they're likely to arrive at the sender in the same IP frame, so they're processed in the same turn.

In the non-tab completion case, the mailbox server gets the OPEN and the ADD(PAKE) in one turn, then gets the RELEASE and the ADD(VERSION) in a subsequent turn, so the two ADDs will go out in separate IP frames, so they'll arrive on the sender in separate frames, so they'll be processed in separate turns, allowing the PAKE to be processed (and send VERSION) before it gets blocked by the processing of VERSION, which allows the receiver to display the Verifier.

**Comment by warner** [comment]
In the tab-completion case, the receiver calls `input_helper.choose_nameplate()` in one turn (when TAB is hit), and  and `choose_words()` much later (after ENTER). 

In the non-tab-completion case, both `choose_nameplate()` and `choose_words()` are called in the same turn.

**Comment by warner** [comment]
Ok, I think I understand it.

In the tab-completion case, the receiver learns the nameplate well before learning the full code. As soon as they learn the nameplate, they'll charge ahead and send CLAIM, then after a roundtrip they get a CLAIMED response so they send OPEN, and a roundtrip after that they get the sender's PAKE. But they don't know the full code yet, so the sender's PAKE gets queued (the `Key` machine goes to the `S01: pake` state). Later, when the user hits ENTER, the full code is known, and the `Key` machine processes the code and the inbound PAKE on the same turn. Processing the code causes the receiver's outbound PAKE message to be put on the transmission queue, and then processing the inbound PAKE lets it compute the unverified key and put the VERSIONS message on the transmission queue. Because these both happened during the same turn, the two outbound messages get merged into the same IP packet, and the mailbox server gets them together, so the sender gets them together.

In the non-tab-completion case, the receiver learns the code (including the nameplate) first. It sends CLAIM for the nameplate and waits for the server's CLAIMED response. When that arrives, it sends both the OPEN and its own PAKE message in the same turn, and waits for the sender's PAKE, which can't arrive until the OPEN is processed. This forces a receiver-server roundtrip before the receiver can submit their VERSIONS message, so the server sees the receiver's PAKE and VERSIONS in separate turns, so they'll wind up on the sender in separate turns.

Ok, that makes me feel better. There's no problem with messages getting merged into a single packet, as long as the receiver doesn't do something silly like make a blocking call to `input()` after sending messages that we expect to be delivered. It's awfully convenient to use `input()` for the sender's "verify?" message, but we can mitigate the blocking nature by inserting that `callLater` delay first, giving any outbound messages a chance to escape.

I'll land that fix shortly. It only changes the behavior of the sender, however, so if you're running a new receiver but an old sender, it will still exhibit the bug (the receiver fails to show the Verifier until after the sender has made a decision).

I'll open a separate issue to think about whether the receiver should have a "confirm?" dialog when `--verify` is in effect, or if the permission-to-send is sufficient.

**Comment by warner** [comment]
Hm. Technically we could change the receiver to put an artificial delay in between processing the code and the inbound PAKE on the same turn, so those two messages wouldn't go out together. That would fix the bug both for new-sender-old-receiver and old-sender-new-receiver.

But, inserting a stall would probably break other aspects of the state machines: they're designed to be run synchronously. And the network is nominally allowed to merge data any way it sees fit (as long as ordering is maintained), so, while unlikely, TCP is within its rights to break that approach. Fixing the sender, as the patch will do, is the "correct" fix, not so dependent upon what the network does.
